### PR TITLE
Calling schema2pkg with Bazel instead of sigh

### LINKED ISSUE #4604: Schema Generation is slow
I've been investigating our build process a bit today and have discovered an opportunity for improvement. 

I removed the `--quiet` flag on the `sigh_command` Bazel rule, added `-x` to the calling bash script, and proceeded to compile all of our Kotlin Tutorial particles. Every time a schema is generated, something like the following output is generated: 

```
INFO: Analyzed target //particles/Tutorial/Kotlin/5_Collections:collections_schemas (96 packages loaded, 37490 targets configured).
...
From Generating Kotlin entity schemas [for host]:
+ set -e
+ cd -P tools/..
++ dirname bazel-out/host/bin/particles/Tutorial/Kotlin/1_HelloWorld/HelloWorld_GeneratedSchemas.wasm.kt
++ basename bazel-out/host/bin/particles/Tutorial/Kotlin/1_HelloWorld/HelloWorld_GeneratedSchemas.wasm.kt
+ tools/sigh schema2wasm --kotlin --wasm --outdir bazel-out/host/bin/particles/Tutorial/Kotlin/1_HelloWorld --outfile HelloWorld_GeneratedSchemas.wasm.kt --package arcs.tutorials particles/Tutorial/Kotlin/1_HelloWorld/HelloWorld.arcs
Warning when linking 0 file(s) already existed.
😌 schema2wasm
🙋 peg
Skipping step; 'src/gen/runtime/manifest-parser.ts' is up-to-date
🙆 peg
🙋 build 
Version 3.7.2

Files:           701
Lines:        174598
Nodes:        719379
Identifiers:  251894
Symbols:      113560
Types:            77
Memory used: 222493K
I/O read:      0.58s
I/O write:     0.00s
Parse time:    7.95s
Bind time:     3.75s
Check time:    0.00s
Emit time:     0.00s
Total time:   11.69s

🙆 build 
🙋 runFn
bazel-out/host/bin/particles/Tutorial/Kotlin/1_HelloWorld/HelloWorld_GeneratedSchemas.wasm.kt
🙆 runFn
🎉 SUCCESS
```

- Around 8-12 seconds of every schema generation is spent running the `build` step in `tools/sigh` -- specifically on the tsc compilation
- This build step is repeated for every schema that is generated. 
- There is relatively negligible time spent in the schema2wasm script portion

This suggests to me that we could get a significant speedup in our build times if we minimize our reliance on tools/sigh.

### PR DESCRIPTION
**Pull request by alxrsngrtn**
This PR provides a refresh of our build process. Instead of relying on `tools/sigh` to run `schema2wasm`, we can use Bazel directly. 

This change will have the following impacts: 
- It will reduce susceptibility of the build to flakes. Every once in a while, the build would fail due to a file in typescript not being there. The root cause of this is having two artifact accounting systems (sigh + Bazel) that don't have full knowledge of each other. This class of flakes will be avoided by relying only on Bazel. 
- Improve the speed of code-generation. This fixes #4604 by not having to re-build the TS project on every call to the script. 
- This allows us to create custom providers for our build process in general. This will be important in the future for compiling policies at APK-build time. 

## Summary of changes
- rules_nodesjs (and TS) support in WORKSPACE
  - To setup properly, this required making modifications to config files and dependencies. I also had to disable the devtools/postinstall script since it was breaking Bazel's installation (relevant bug: #5379). 
- Created a `arcs_ts_preprocessing` rule that coalesces imports to use the `-node.js` platform
- Copied / Converted `xen-state` from modalities to a TS version in runtime. 
- Created a `schema2pkg` Bazel rule + nodejs binary that is called from the `arcs_kt_schema` macro. 

## Build Performance (WIP)
Tested on a my work MacBook pro with the standard project setup. I performed the following command (after `bazel clean`) for this branch and master: `bazel build //java/... //javatests/... //src/... //particles/...`
- master: `INFO: Elapsed time: 819.224s, Critical Path: 373.69s`
- this branch: `INFO: Elapsed time: 671.487s, Critical Path: 415.46s`

### CODE REVIEW

File: java/arcs/core/data/testdata/BUILD [hunk 1]
```diff
@@ -16,16 +17,21 @@ filegroup(
...
     name = "import_example_gen",
     srcs = ["ImportExample.arcs"],
     deps = ["Schemas.arcs"],
 )
 
-arcs_kt_gen(
-    name = "example_gen",
-    srcs = ["WriterReaderExample.arcs"],
-)
+#arcs_kt_gen(
```
Comment (cromwellian) [inline java/arcs/core/data/testdata/BUILD:31]: remove?

File: package.json [hunk 2]
```diff
@@ -9,7 +9,6 @@
   "module": "dist/webmain.js",
   "main": "dist/index.es.js",
   "scripts": {
-    "prepare": "cross-env tools/sigh check && cd devtools && npm i && cd ../concrete-storage && npm i",
```
Comment (cromwellian) [inline package.json:12]: you could work around this by having a separate package.json for Bazel I guess

Comment (alxrsngrtn) [reply package.json]: I tried this; it didn't work. Somewhere the rules_nodejs package looks for a file called `package.json` in the root directory, despite the argument that one supplies for the package file. 

I did find out what the problem was and filed a bug for it #5379.

**Comment by alxrsngrtn** [comment]
@cromwellian Do you know how I can call peg / generate the manifest parser in these new bazel rules? 
```
bazel-out/host/bin/src/runtime/manifest.ts:11:21 - error TS2307: Cannot find module '../gen/runtime/manifest-parser.js'.

11 import {parse} from '../gen/runtime/manifest-parser.js';
```

**Comment by alxrsngrtn** [comment]
After asking the question above, an idea comes to mind: 
I can create a Bazel rule that calls sigh (at least for now) that runs the peg generator. `runtime_libs` can depend on the output. While sigh is not eliminated, it's invocation is greatly reduced.

**Comment by cromwellian** [comment]
> @cromwellian Do you know how I can call peg / generate the manifest parser in these new bazel rules?
> 
> ```
> bazel-out/host/bin/src/runtime/manifest.ts:11:21 - error TS2307: Cannot find module '../gen/runtime/manifest-parser.js'.
> 
> 11 import {parse} from '../gen/runtime/manifest-parser.js';
> ```

You can try to port my gentspegjs, see third_party/javascript/arcs/BUILD

File: WORKSPACE [hunk 3]
```diff
@@ -4,18 +4,26 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
...
-    sha256 = "0942d188f4d0de6ddb743b9f6642a26ce1ad89f09c0035a9a5ca5ba9615c96aa",
-    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/0.38.1/rules_nodejs-0.38.1.tar.gz"],
+    sha256 = "f9e7b9f42ae202cc2d2ce6d698ccb49a9f7f7ea572a78fd451696d03ef2ee116",
+    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/1.6.0/rules_nodejs-1.6.0.tar.gz"],
 )
 
-load("@build_bazel_rules_nodejs//:index.bzl", "node_repositories")
+load("@build_bazel_rules_nodejs//:index.bzl", "yarn_install")
 
-node_repositories(
```
Comment (cromwellian) [inline WORKSPACE:13]: why remove this? Shouldn't we still fix the yarn/node versions used?

File: third_party/java/arcs/build_defs/internal/kotlin.bzl [hunk 4]
```diff
@@ -206,7 +206,7 @@ def arcs_kt_library(
             testonly = testonly,
             # Exclude any wasm-specific srcs.
             # TODO: jvm srcs will be included here. That is not what we want.
-            srcs = [src for src in srcs if not src.endswith(".wasm.kt")],
+            srcs = [src for src in srcs if src.find("_genrule_wasm") == -1],
```
Comment (cromwellian) [inline third_party/java/arcs/build_defs/internal/kotlin.bzl:209]: I fill like these magic suffixes should be pulled into constants, or a function like
is_src_for_platform(src, "js")

File: third_party/java/arcs/build_defs/internal/kotlin.bzl [hunk 5]
```diff
@@ -445,6 +451,7 @@ def arcs_kt_plan(name, srcs = [], data = [], deps = [], platforms = ["jvm"], vis
       data: list of Arcs manifests needed at runtime
```
Comment (cromwellian) [inline third_party/java/arcs/build_defs/internal/kotlin.bzl:451]: eventually we should convert these to skylark rules so we can use providers to roll up the transitive deps and avoid suffix magic and data vs deps tricks.

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 6]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
...
+    ]
+
+    ctx.actions.run_shell(
+        inputs = ctx.files.srcs,
+        outputs = outputs,
+        arguments = [src.path for src in ctx.files.srcs],
+        command = """
+        for i in "$@"; do
+          out="$(pwd)/bazel-out/host/bin/$i"
+          sed -e 's/-web.js/-node.js/g' $i > "$out"
```
Comment (cromwellian) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:175]: eventually we will want to build the entire web version in bazel as well, perhaps there should be an enable/disable flag (default platform, node)

File: .bazelignore [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # Some npm modules include their own WORKSPACE/BUILD files, so we need to ignore
 # them from our build. Manually listing them is not ideal, but we can't ignore
 # the entire node_modules folder, because we need it for our bazel builds.
```
Comment (piotrswigon) [inline .bazelignore:3]: Comment is stale, please update

File: devtools/package.json [hunk 8]
```diff
@@ -23,7 +23,6 @@
     "js-green-licenses": "1.1.0"
   },
   "scripts": {
-    "postinstall": "cross-env empathy install -a deps && cross-env tools/css-module-wrap ./node_modules/vis/dist/vis-timeline-graph2d.min.css ./node_modules/golden-layout/src/css/goldenlayout-base.css ./node_modules/golden-layout/src/css/goldenlayout-light-theme.css",
```
Comment (piotrswigon) [inline devtools/package.json:26]: Why are you removing this, this performed certain functions that are needed. Why is this removal necessary?

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 9]
```diff
@@ -112,50 +114,43 @@ def arcs_kt_schema(
...
 
-    if (test_harness):
+    if test_harness:
         test_harness_outs = []
         for src in srcs:
-            out = replace_arcs_suffix(src, "_TestHarness.kt")
+            out = replace_arcs_suffix(src, "_genrule_test_harness")
             test_harness_outs.append(out)
 
-            _run_schema2wasm(
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:144]: Can you remove the definition of _run_schema2wasm now?

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 10]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
         )
     return {"outs": outs, "deps": outdeps}
 
+def _arcs_ts_preproc_impl(ctx):
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:162]: A comment would be useful.

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 11]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
...
+    )
+
+    return DefaultInfo(files = depset(outputs))
+
+arcs_ts_preprocessing = rule(
+    implementation = _arcs_ts_preproc_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = [".ts", ".d.ts"]),
+    },
+    doc = """Converts to node platforms for the arcs ts runtime.""",
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:187]: A more elaborate documentation would be useful.

Maybe something along the lines of:

```
Processes Arcs TypeScript sources to refer to node variants of the dependencies in the platform/ directory.
```

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 12]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
...
+)
+
+def _schema2pkg_impl(ctx):
+    output_name = ctx.label.name + "_GeneratedSchemas.%s.kt" % ctx.attr.platform
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:191]: if ctx.attr.platform == "cpp" would we get `_GeneratedSchemas.cpp.kt` ?

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 13]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
...
+            doc = """whether to output a particle test harness only (requires lang=kotlin)""",
+        ),
+        "_compiler": attr.label(
+            cfg = "host",
+            default = Label("//src/tools:schema2pkg"),
+            allow_files = True,
+            executable = True,
+        ),
+    },
+    doc = """Stand-alone schema2* tool.""",
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:250]: The fact that it is standalone is only relevant for this PR. Once you land this PR, there will be no other way than stand-alone, so it's not really relevant. When writing docs you should think in terms of what makes sense to convey to the reader who will be reading this code tomorrow and a year from now. Typically you should not convey the history of what code went through, unless it is important in understanding what the code does - I don't think it is the case here.

Given this is now *the* rule for schema2pkg, you should write a more proper documentation for it, at least explain the arguments, usage etc.

File: third_party/java/arcs/build_defs/internal/schemas.bzl [hunk 14]
```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
...
         srcs,
         data = [],
         deps = [],
         platforms = ["jvm"],
         test_harness = True,
+        testonly=False,
         visibility = None):
     """Generates Kotlin files for the given .arcs files.
 
+    TODO(alxr): Move to manifest.bzl file
```
Comment (piotrswigon) [inline third_party/java/arcs/build_defs/internal/schemas.bzl:264]: If there is a slightest chance that this TODO will still be here 3 months from now then I would restrain from adding it.

Project codebase is not a notebook/to-do list for the developers, unless a TODO reveals something important about how the code is structured or how it works it is not a great idea to decorate it with notes on how we would like it to look like.

File: src/runtime/xen-state.ts [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+// tslint:disable:no-any
+
+const nob = () => Object.create(null);
+
+// @ts-ignore
+const debounce = (key, action, delay) => {
```
Comment (piotrswigon) [inline src/runtime/xen-state.ts:15]: Does this mean we will have now 2 divergent versions of xen-state? If so, is there any way to avoid it? E.g. adding .d.ts file instead of making a full on copy?

File: .bazelignore (hunk 7 above)
Comment (alxrsngrtn) [reply .bazelignore:3]: Thanks!

File: WORKSPACE (hunk 3 above)
Comment (alxrsngrtn) [reply WORKSPACE:13]: This rule is states as an alternative option to having modules be managed by yarn ([source](https://bazelbuild.github.io/rules_nodejs/Built-ins.html#creating-dependency-installation-scripts-for-manually-managed-dependencies)).

After a quick scan of the documentation, there doesn't seem to be an option to specify the node version in `yarn_install`. I imagine that this could be controlled from the `package.json`? I'll keep exploring.

File: src/runtime/xen-state.ts (hunk 15 above)
Comment (alxrsngrtn) [reply src/runtime/xen-state.ts:15]: In the OSS world, there's no way to have a mixture of typescript and javascript in a library. I tried to find a way around this, but this was the only way that seemed trackable (I didn't want to create custom bazel rules to handle this case). So, there wasn't an option to add a `.d.ts` file. :(

File: third_party/java/arcs/build_defs/internal/kotlin.bzl (hunk 5 above)
Comment (alxrsngrtn) [reply third_party/java/arcs/build_defs/internal/kotlin.bzl:452]: I totally agree. For now, I didn't want to break any interfaces.

File: third_party/java/arcs/build_defs/internal/schemas.bzl (hunk 9 above)
Comment (alxrsngrtn) [reply third_party/java/arcs/build_defs/internal/schemas.bzl:144]: This was a good reminder: I should change the cpp side of the rules (that's the last place where this fn is used).

**Comment by alxrsngrtn** [comment]
Closing for now, will reconsider moving forward with this at a future time.
